packages/dp-auctions-lib/src/exec-transactions.js

```diff
@@ -0,0 +1,202 @@
+'use strict'
+
+const { EventEmitter } = require('events')
+const Big = require('big.js').default
+const debug = require('debug')('vesper-lib:exec-txs')
+const pSeries = require('p-series')
+const pTap = require('p-tap').default
+
+const fromUnit = (number, decimals = 18) =>
+  Big(`${Big(number).toFixed()}e-${decimals}`).toFixed()
+
+const calculateFee = ({ transaction, receipt }) =>
+  Big(transaction.gasPrice).times(receipt.gasUsed).toFixed()
+
+const calculateTotalFee = (transactionsData) =>
+  transactionsData
+    .map(calculateFee)
+    .reduce((total, fee) => Big(total).plus(fee), Big(0))
+    .toFixed()
+
+const createEstimateGasAndSend = (web3, emitter, overestimation = 1.25) =>
+  function (method, transactionOptions, suffix) {
+    const suffixed = (event) => `${event}${suffix ? `-${suffix}` : ''}`
+
+    let hash
+    let transactionPromise
+
+    const estimateGas = function () {
+      debug('Estimating gas')
+
+      const estimationPromise = method
+        .estimateGas(transactionOptions)
+        .then(
+          pTap(function (gas) {
+            debug('Gas needed is %d (x%s)', gas, overestimation.toFixed(2))
+          })
+        )
+        .then((gas) => Math.ceil(gas * overestimation))
+        .then(function (gas) {
+          // Emit the result
+          emitter.emit(suffixed('estimatedGas'), gas)
+          return gas
+        })
+
+      estimationPromise.catch(function (err) {
+        debug('Gas estimation failed: %s', err.message)
+        if (!emitter.listenerCount('error')) {
+          return
+        }
+        emitter.emit('error', err)
+      })
```

Comment:
it might be a silly question, but why do you need a variable for the promise? the only reason I could think of is that you want to allow the consumer to catch the error and still receive the `.on('error)` - is that correct?

Reply:
Correct. Or allow the user to listen for the `error` event, don't use a catch and prevent an "uncaught" error.